common/lib/client/realtimepresence.js

```diff
@@ -9,6 +9,15 @@ var RealtimePresence = (function() {
 		return realtimePresence.channel.realtime.auth.clientId;
 	}
 
+	function isAnonymous(realtimePresence) {
+		var realtime = realtimePresence.channel.realtime;
+		/* If not currently connected, we can't assume that we're an anonymous
+		 * client, as realtime may inform us of our clientId in the CONNECTED
+		 * message. So assume we're not anonymous and leave it to realtime to
+		 * return an error if we are */
+		return !realtime.auth.clientId && realtime.connection.state === 'connected';
```

Comment:
> If for example, a user has provided a clientId in the ClientOptions, but not yet connected, then...

...then `isAnonymous` will (correctly) return false. If someone specifies `auth.clientId` in clientOptions, then [auth.clientId will be set on client lib init](https://github.com/ably/ably-js/blob/e6319f9dc8d5ea802cdbfb6b24b09175231f7300/common/lib/client/auth.js#L47-L50), so `!realtime.auth.clientId` will be false. (so the function doesn't bother checking the connection state due to short-circuiting)

The function might return false even if the user will end up being anonymous after all (which is fine, as then realtime can reject the ENTER), but it should never return true when the user might not be. (I originally called it `isDefinitelyAnonymous` to make that clear, but Paddy requested `isAnonymous`, see other thread)